File: cpp/oop/diamondInheritance/Bird.cpp
#include <iostream>
#include <string>

#include "Bird.h"

using namespace std;

Bird::Bird(string name) :
    Animal(name)
{
    cout << "Bird::Bird: name=" << name << endl;
}

Bird::~Bird() {
    cout << "Bird::~Bird" << endl;
}

void Bird::fly() {
    cout << "Bird::fly" << endl;
}

void Bird::run() {
    cout << "Bird::run: 2 legs"  << endl;
}

File: java/oop/diamondInheritance/Bird.java
public class Bird extends Animal implements ICanFlyAnimal {
    public Bird(String name) {
	super(name);
	System.out.println("[D] Bird::Bird: name=[" + name + "]");
    }

    public void fly() {
	System.out.println("[D] Bird::fly with 2 wings and 2 legs");
    }

    public void run() {
	System.out.println("[D] Bird::run with 2 legs and 2 wings");
    }
}

